internal/docstore/docstore.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2019 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package docstore provides a portable implementation of a document store.
+package docstore // import "gocloud.dev/internal/docstore"
+
+import (
+	"context"
+	"strings"
+	"unicode/utf8"
+
+	"gocloud.dev/internal/docstore/driver"
+	"gocloud.dev/internal/gcerr"
+)
+
+// A Document is a set of field-value pairs.
+// It can be represented as a map[string]int or a pointer to a struct.
+// For structs, the exported fields are the document fields.
+type Document = interface{}
+
+// A Collection is a set of documents.
+type Collection struct {
+	driver driver.Collection
+}
+
+// NewCollection is intended for use by provider implementations.
+var NewCollection = newCollection
+
+// newCollection makes a Collection.
+func newCollection(d driver.Collection) *Collection {
+	return &Collection{driver: d}
+}
+
+// Actions returns an ActionList that can be used to perform
+// actions on the collection's documents.
+func (c *Collection) Actions() *ActionList {
+	return &ActionList{coll: c}
+}
+
+// An ActionList is a sequence of actions that affect a single collection.
```

Comment:
Describe the semantics -- they are performed in order (not in parallel), and an error on one means the others aren't performed? Or are they?

Reply:
Done.